source/agora/node/Ledger.d

```diff
@@ -2178,7 +2196,45 @@ unittest
     scope ledger = new TestLedger(genesis_validator_keys[0], blocks, params, 600.seconds, mock_clock);
 
     // Add preimages for all validators (except for two of them) till end of cycle
-    auto skip_indexes = [ 2, 5 ];
+    uint[] skip_indexes = [ 2, 5 ];
+
+    ledger.simulatePreimages(Height(params.ValidatorCycle), skip_indexes);
+
+    // Block with no fee
+    auto no_fee_txs = blocks[$-1].spendable.map!(txb => txb.sign()).array();
+    no_fee_txs.each!(tx => assert(ledger.acceptTransaction(tx)));
+
+    ConsensusData data;
+    ledger.prepareNominatingSet(data, Block.TxsInTestBlock, mock_clock.networkTime());
+
+    assert(ledger.validateConsensusData(data, skip_indexes) is null);
+
+    data.missing_validators = [2,5,7];
+    assert(ledger.validateConsensusData(data, [2,3,5,7,9]) is null);
+
+    data.missing_validators = [2,5];
+    assert(ledger.validateConsensusData(data, [2]) == "Higher bound violation - Missing validator mismatch [2, 5] is not a subset of [2]");
+
+    data.missing_validators = [5];
+    assert(ledger.validateConsensusData(data, [2]) == "Lower bound violation - Missing validator mismatch [2, 5] is not a subset of [5]");
```

Comment:
There are four `validateConsensusData` calls so it would be good to explain the concept of each call.